nodecg-io-serial/extension/SerialClient.ts

```diff
@@ -0,0 +1,95 @@
+import { ServiceClient } from "nodecg-io-core/extension/types";
+import { success, error, Result } from "nodecg-io-core/extension/utils/result";
+import * as SerialPort from "serialport"; // This is neccesary, because serialport only likes require!
+
+export interface DeviceInfo {
+    port: string;
+    manucaturer: string;
+    serialNumber: string;
+    pnpId: string;
+}
+
+interface Protocol {
+    delimiter: "\n\r" | "\n";
+    encoding: "ascii" | "utf8" | "utf16le" | "ucs2" | "base64" | "binary" | "hex" | undefined;
+}
+
+export interface SerialServiceConfig {
+    device: DeviceInfo;
+    connection: SerialPort.OpenOptions;
+    protocol: Protocol;
+}
+
+export class SerialServiceClient implements ServiceClient<SerialPort> {
+    private serialPort: SerialPort;
+    private parser: SerialPort.parsers.Readline;
+    constructor(private readonly settings: SerialServiceConfig) {}
+
+    async init(): Promise<void> {
+        const port = await SerialServiceClient.inferPort(this.settings.device);
+        if (!port.failed) {
+            this.serialPort = new SerialPort(port.result, this.settings.connection);
+            this.parser = this.serialPort.pipe(new SerialPort.parsers.Readline(this.settings.protocol));
+        } else {
+            console.log("Could not initialize the serial port");
+            // TODO: handle error properly
+        }
+    }
+
+    getNativeClient(): SerialPort {
+        return this.serialPort;
+    }
+
+    static async inferPort(deviceInfo: DeviceInfo): Promise<Result<string>> {
+        const result = [];
+        const devices = await SerialPort.list();
+        if (deviceInfo.port) {
+            result.push(deviceInfo.port);
+        } else {
+            devices.forEach((element) => {
+                if (deviceInfo.pnpId && "pnpId" in element && element.pnpId === deviceInfo.pnpId) {
+                    result.push(element["path"]);
+                } else if (
+                    deviceInfo.manucaturer &&
+                    deviceInfo.serialNumber &&
+                    element.manufacturer === deviceInfo.manucaturer &&
+                    element.serialNumber === deviceInfo.serialNumber
+                ) {
+                    result.push(element["path"]);
+                }
```

Comment:
nit: use direct accessing of properties instead of strings because they can be type checked.
```suggestion
                if (deviceInfo.pnpId && element.pnpId && element.pnpId === deviceInfo.pnpId) {
                    result.push(element.path);
                } else if (
                    deviceInfo.manucaturer &&
                    deviceInfo.serialNumber &&
                    element.manufacturer === deviceInfo.manucaturer &&
                    element.serialNumber === deviceInfo.serialNumber
                ) {
                    result.push(element.path);
                }
```

Reply:
I tried it, but the compiler said that it the attribute does not exist and it can not infer the type of element. Therefore it tells me that element has an implicit type of any....
I don't really know how to fix that in a clean way besides importing the types of serialport directly.

But I totally agree with you that it should be like that!

Reply:
nvm.... I tried it again and it works fine